pylint/plugins/hass_enforce_type_hints.py

```diff
@@ -0,0 +1,224 @@
+        module_filter=_MODULE_FILTERS["init"],
+        function_name="async_setup_entry",
+        arg_types={
+            0: _ARGUMENT_MATCH["HomeAssistant"],
+            1: _ARGUMENT_MATCH["ConfigEntry"],
+        },
+        return_type=_RETURN_MATCH["bool"],
+    ),
+    TypeHintMatch(
+        module_filter=_MODULE_FILTERS["init"],
+        function_name="async_remove_entry",
+        arg_types={
+            0: _ARGUMENT_MATCH["HomeAssistant"],
+            1: _ARGUMENT_MATCH["ConfigEntry"],
+        },
+        return_type=_RETURN_MATCH["None"],
+    ),
+    TypeHintMatch(
+        module_filter=_MODULE_FILTERS["init"],
+        function_name="async_unload_entry",
+        arg_types={
+            0: _ARGUMENT_MATCH["HomeAssistant"],
+            1: _ARGUMENT_MATCH["ConfigEntry"],
+        },
+        return_type=_RETURN_MATCH["bool"],
+    ),
+    TypeHintMatch(
+        module_filter=_MODULE_FILTERS["init"],
+        function_name="async_migrate_entry",
+        arg_types={
+            0: _ARGUMENT_MATCH["HomeAssistant"],
+            1: _ARGUMENT_MATCH["ConfigEntry"],
+        },
+        return_type=_RETURN_MATCH["bool"],
+    ),
+]
+
+
+def _get_all_annotation(node: astroid.FunctionDef) -> list[astroid.NodeNG | None]:
+    args = node.args
+    annotations = (
+        args.posonlyargs_annotations + args.annotations + args.kwonlyargs_annotations
+    )
+    if args.vararg is not None:
+        annotations.append(args.varargannotation)
+    if args.kwarg is not None:
+        annotations.append(args.kwargannotation)
+    return annotations
+
+
+def _has_valid_annotations(
+    annotations: list[astroid.NodeNG | None],
+) -> list[astroid.NodeNG | None]:
+    for annotation in annotations:
+        if annotation is not None:
+            return True
+    return False
+
+
+class HassTypeHintChecker(BaseChecker):  # type: ignore[misc]
+    """Checker for setup type hints."""
+
+    __implements__ = IAstroidChecker
+
+    name = "hass_enforce_type_hints"
+    priority = -1
+    msgs = {
+        "W0020": (
+            "Argument should be of type HomeAssistant",
+            "hass-type-hint-hass",
+            "Used when method argument isn't HomeAssistant as expected",
+        ),
+        "W0021": (
+            "Argument should be of type ConfigType",
+            "hass-type-hint-config-type",
+            "Used when method argument isn't ConfigType as expected",
+        ),
+        "W0022": (
+            "Argument should be of type ConfigEntry",
+            "hass-type-hint-config-entry",
+            "Used when method argument isn't ConfigEntry as expected",
+        ),
+        "W0023": (
+            "Return type should be of type bool",
+            "hass-type-hint-return-bool",
+            "Used when method return type isn't bool",
+        ),
+        "W0024": (
+            "Return type should be None",
+            "hass-type-hint-return-none",
+            "Used when method return type isn't None",
+        ),
+    }
```

Comment:
This doesn't really scale if we will have more such checks. You can have generic warnings instead.
```python3
        "W0020": (
            "Argument %d type should be %s",
            "hass-argument-type",
            "Used when method argument type is incorrect",
        ),
       "W0021": (
            "Return type should be %s",
            "hass-return-type",
            "Used when method return type is incorrect",
        ),
```

Reply:
That would be good, but how do I pass `%s` and `%d`?
```
homeassistant/components/rfxtrx/__init__.py:84:34: W0020: Argument %d type should be %s (hass-argument-type)
homeassistant/components/rfxtrx/__init__.py:102:35: W0020: Argument %d type should be %s (hass-argument-type)
```

Reply:
I found it - I need to add args to add_message function:
```python
                self.add_message(
                    "hass-argument-type",
                    node=node.args.args[key],
                    args=(key, value.expected_type),
                )
```